File: crates/injector/src/injector.rs
use std::{ffi, mem, ptr};
use sysinfo::*;
use winapi::shared::minwindef::LPVOID;
use winapi::um::errhandlingapi::GetLastError;
use winapi::um::libloaderapi::{GetProcAddress, LoadLibraryA};
use winapi::um::memoryapi::{VirtualAllocEx, VirtualQueryEx, WriteProcessMemory};
use winapi::um::minwinbase::LPTHREAD_START_ROUTINE;
use winapi::um::processthreadsapi::{CreateRemoteThread, OpenProcess};
use winapi::um::psapi::GetModuleBaseNameA;
use winapi::um::synchapi::WaitForSingleObject;
use winapi::um::winbase::INFINITE;
use winapi::um::winnt::MEMORY_BASIC_INFORMATION;
use winapi::um::winnt::{HANDLE, MEM_COMMIT, PROCESS_ALL_ACCESS};

pub struct MemoryMap {
    addr: usize,
    name: String,
}

pub struct Process {
    handle: HANDLE,
    pub pid: Pid,
}

pub unsafe fn memory_map(proc: &Process) -> Vec<MemoryMap> {
    let mut result: Vec<MemoryMap> = vec![];
    let mut cur = 0;
    let mut mbf: MEMORY_BASIC_INFORMATION = mem::zeroed();
    let mut buffer = vec![0u8; 0x1000];

    while VirtualQueryEx(
        proc.handle,
        mem::transmute(cur),
        &mut mbf,
        mem::size_of::<MEMORY_BASIC_INFORMATION>(),
    ) != 0
    {
        let res = GetModuleBaseNameA(
            proc.handle,
            mem::transmute(cur),
            buffer.as_mut_ptr() as *mut i8,
            buffer.len() as u32,
        );
        if res != 0 {
            let name = std::str::from_utf8_unchecked(&buffer[0..res as usize]);
            result.push(MemoryMap {
                addr: cur,
                name: name.to_string(),
            })
        };
        cur += mbf.RegionSize;
    }
    result
}

pub unsafe fn find_base(proc: &Process, name: &str) -> Option<usize> {
    let map = memory_map(proc);
    let mut res = map
        .iter()
        .filter(|item| name.to_lowercase().contains(&item.name.to_lowercase()));

    Some(res.next()?.addr)
}

unsafe fn alloc(proc: &Process, size: usize) -> LPVOID {
    let res = VirtualAllocEx(
        proc.handle,
        ptr::null_mut(),
        (size + 0xFFF) & !0xFFF,
        MEM_COMMIT,
        0x40,
    );
    if res.is_null() {
        panic!(format!("Allocation failed: {:x}", GetLastError()))
    }
    log::debug!("Allocated memory: {:x}", res as usize);
    res
}

unsafe fn alloc_str(proc: &Process, str: &str) -> LPVOID {
    let addr = alloc(proc, str.len() + 1);
    write_mem(proc, addr, str);
    addr
}

unsafe fn write_mem(proc: &Process, addr: LPVOID, str: &str) {
    WriteProcessMemory(
        proc.handle,
        addr,
        str.as_ptr() as LPVOID,
        str.len(),
        ptr::null_mut(),
    );
}

macro_rules! to_cstr {
    ($l:expr) => {
        ffi::CString::new($l).unwrap()
    };
}

pub unsafe fn find_function(
    proc: &Process,
    library: &str,
    function: &str,
) -> LPTHREAD_START_ROUTINE {
    let library_name = to_cstr!(library);
    let function_name = to_cstr!(function);
    let library_ptr = LoadLibraryA(library_name.as_ptr());
    mem::transmute(
        (GetProcAddress(library_ptr, function_name.as_ptr()) as usize)
            .wrapping_sub(library_ptr as usize)
            .wrapping_add(find_base(proc, library).unwrap()),
    )
}

pub unsafe fn inject_dll(proc: &Process, name: &str) {
    let str = alloc_str(proc, name);
    log::debug!("Injecting DLL into process... {}", name);
    call(
        proc,
        find_function(proc, "kernel32.dll", "LoadLibraryA"),
        str,
    );
}

pub unsafe fn call(proc: &Process, addr: LPTHREAD_START_ROUTINE, args: LPVOID) {
    log::info!(
        "Calling: {:x}",
        std::mem::transmute::<LPTHREAD_START_ROUTINE, usize>(addr)
    );
    let handle = CreateRemoteThread(
        proc.handle,
        ptr::null_mut(),
        0,
        addr,
        args,
        0,
        ptr::null_mut(),
    );
    WaitForSingleObject(handle, INFINITE);
}

pub unsafe fn find_process(name: &str) -> Option<Process> {
    let mut system = sysinfo::System::new_all();
    log::debug!("Refreshing the process list...");
    system.refresh_processes();

    log::debug!(
        "Iterating through {} processes...",
        system.get_processes().len()
    );

    for (pid, proc_) in system.get_processes() {
        if proc_.name().to_lowercase() == name.to_lowercase() {
            let handle = OpenProcess(PROCESS_ALL_ACCESS, 0, *pid as u32);
            return Some(Process { handle, pid: *pid });
        }
    }

    None
}

File: src/injector/injector.cpp
#include "injector.h"

#include <Psapi.h>     // for GetModuleBaseNameA
#include <TlHelp32.h>  // for PROCESSENTRY32, CreateToolhelp32Snapshot, Pro...
#include <algorithm>   // for transform
#include <cctype>      // for tolower
#include <cstring>     // for size_t, strrchr, NULL
#include <new>         // for operator new
#include <type_traits> // for move, forward
#include <utility>     // for min
#include <vector>      // for vector, _Vector_iterator, _Vector_const_iterator

#include "logger.h" // for PANIC, DEBUG

using namespace std::chrono_literals;

std::vector<MemoryMap> memory_map(const Process& proc)
{
    std::vector<MemoryMap> result;
    size_t cur = 0;
    MEMORY_BASIC_INFORMATION mbf = {};
    char buffer[0x1000] = {};

    while (VirtualQueryEx(proc.handle, (LPVOID)cur, &mbf, sizeof(mbf)) != 0)
    {
        auto res = GetModuleBaseNameA(proc.handle, (HMODULE)cur, buffer, sizeof(buffer));
        if (res != 0)
        {
            result.push_back(MemoryMap{cur, std::string(buffer)});
        };
        cur += mbf.RegionSize;
    }
    return result;
}

std::string to_lower(std::string s)
{
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c)
                   { return (unsigned char)std::tolower(c); });
    return s;
}

size_t find_base(Process proc, std::string name)
{
    auto map = memory_map(proc);
    name = to_lower(name);
    for (auto&& item : map)
    {
        if (name.find(to_lower(item.name)) != std::string::npos)
            return item.addr;
    }

    ERR("Cannot find library in the target process: {}", name);
    return 0;
}

LPVOID alloc(Process proc, size_t size)
{
    auto res = VirtualAllocEx(proc.handle, nullptr, (size + 0xFFF) & ~0xFFF, MEM_COMMIT, 0x40);
    if (res == NULL)
    {
        PANIC("Allocation failed: {:#x}", GetLastError());
    }
    // DEBUG("Allocated memory: {}", res);
    return res;
}

void write_mem(Process proc, LPVOID addr, std::string str)
{
    WriteProcessMemory(proc.handle, addr, str.data(), str.size(), NULL);
}

LPVOID alloc_str(Process proc, std::string str)
{
    auto addr = alloc(proc, str.size() + 1);
    write_mem(proc, addr, str);
    return addr;
}

LPTHREAD_START_ROUTINE find_function(const Process& proc, const std::string& library, const std::string& function)
{
    auto library_ptr = (size_t)LoadLibraryA(library.data());

    if (library_ptr == 0)
    {
        ERR("Cannot find the address of the library in current process: {}", library.data());
        return nullptr;
    }

    auto addr = (size_t)GetProcAddress((HMODULE)library_ptr, function.data());

    if (addr == 0)
    {
        ERR(
            "Cannot find the address of the function in current process: {} :: "
            "{}",
            library.data(),
            function.data());
        return nullptr;
    }

    auto base = find_base(proc, library);
    if (base == 0)
        return nullptr;

    return reinterpret_cast<LPTHREAD_START_ROUTINE>(addr - library_ptr + base);
}

void call(const Process& proc, LPTHREAD_START_ROUTINE addr, LPVOID args)
{
    // DEBUG("Calling: {}", (void*)addr);
    auto handle = CreateRemoteThread(proc.handle, nullptr, 0, addr, args, 0, nullptr);
    WaitForSingleObject(handle, INFINITE);
}

bool find_dll_in_process(DWORD pid, const std::string& name)
{
    HMODULE hMods[1024];
    HANDLE hProcess;
    DWORD cbNeeded;
    unsigned int i;
    hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, pid);
    if (NULL == hProcess)
        return false;
    if (EnumProcessModules(hProcess, hMods, sizeof(hMods), &cbNeeded))
    {
        for (i = 0; i < (cbNeeded / sizeof(HMODULE)); i++)
        {
            TCHAR szModName[MAX_PATH];
            if (GetModuleFileNameEx(hProcess, hMods[i], szModName, sizeof(szModName) / sizeof(TCHAR)))
            {
                auto modName = std::string(szModName);
                if (modName.ends_with(name))
                    return true;
            }
        }
    }
    return false;
}

void inject_dll(const Process& proc, const std::string& name)
{
    auto str = alloc_str(proc, name);
    INFO("Injecting DLL into process... {}", name);
    int tries = 1;
    LPTHREAD_START_ROUTINE func;
    do
    {
        func = find_function(proc, "KERNEL32.DLL", "LoadLibraryA");
        if (func == nullptr)
        {
            if (tries < 4)
            {
                INFO("Retrying in a jiffy...");
                std::this_thread::sleep_for(tries * 1s);
            }
        }
    } while (func == nullptr && ++tries < 5);
    if (func == nullptr)
        PANIC("Injecting failed, maybe you should try --launch_game instead...");
    call(proc, func, str);
}

std::vector<ProcessInfo> get_processes()
{
    // No unicode
#undef Process32First
#undef Process32Next
#undef PROCESSENTRY32
    std::vector<ProcessInfo> res;
    auto snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (snapshot == nullptr)
        return {};

    PROCESSENTRY32 ppe = {sizeof(ppe)};
    auto proc = Process32First(snapshot, &ppe);

    while (proc)
    {
        auto name = ppe.szExeFile;
        if (auto delim = strrchr(name, '\\'))
            name = delim;
        res.push_back({name, ppe.th32ProcessID});
        proc = Process32Next(snapshot, &ppe);
    }
    return res;
}

std::optional<Process> find_process(std::string name)
{
    for (auto& proc : get_processes())
    {
        if (proc.name == name)
        {
            return Process{OpenProcess(PROCESS_ALL_ACCESS, 0, proc.pid), proc};
        }
    }
    return {};
}
